python/ccxt/base/exchange.py

```diff
@@ -1093,31 +1102,12 @@ def load_accounts(self, reload=False, params={}):
         self.accountsById = self.index_by(self.accounts, 'id')
         return self.accounts
 
-    def populate_fees(self):
-        if not (hasattr(self, 'markets') or hasattr(self, 'currencies')):
-            return
-
-        for currency, data in self.currencies.items():  # try load withdrawal fees from currencies
-            if 'fee' in data and data['fee'] is not None:
-                self.fees['funding']['withdraw'][currency] = data['fee']
-                self.fees['funding']['fee_loaded'] = True
-
-        # find a way to populate trading fees from markets
-
-    def load_fees(self):
+    def load_fees(self, reload=False):
         self.load_markets()
-        self.populate_fees()
-        if not (self.has['fetchTradingFees'] or self.has['fetchFundingFees']):
-            return self.fees
-
-        fetched_fees = self.fetch_fees()
-        if fetched_fees['funding']:
-            self.fees['funding']['fee_loaded'] = True
-        if fetched_fees['trading']:
-            self.fees['trading']['fee_loaded'] = True
-
-        self.fees = self.deep_extend(self.fees, fetched_fees)
-        return self.fees
+        if not reload:
+            if self.loaded_fees == Exchange.loaded_fees:
+                self.loaded_fees = self.deep_extend(self.loaded_fees, self.fetch_fees())
```

Comment:
One thing that bothers me with this approach is that some exchanges don't have an endpoint for fetching all fees at once – so, fees are market-specific or currency-specific. And those exchanges don't have `fetch_trading_fees (params)`, but have `fetch_trading_fee (symbol, params)`. Also, the base implementation of `fetch_trading_fee` (for one symbol) is trying to `fetch_trading_fees()` (all at once) and take the fee from there.

I'd prefer not storing the fees (but the hardcoded fees) on the exchange instance – i think a better approach would be to store them on the corresponding markets... So, maybe, we need 

```Python
def load_trading_fee(self, symbol, reload=False):
    self.load_markets()
    if reload:
        if not self.markets[symbol]['loadedFee']:
            fee = self.fetch_trading_fee (symbol)
            self.markets[symbol]['maker'] = fee['maker']
            self.markets[symbol]['taker'] = fee['taker']
            self.markets[symbol]['loadedFee'] = True 
    return self.markets[symbol]
```

↑ From there it follows that we may also need to add a `fee` property on the market and make `taker` and `maker` subproperties of `market[symbol]['fee'] = { 'taker': ..., 'maker': ..., otherFlags..., loaded, notLoaded, unloaded, reloaded, etc... }`

What do you think on the above?

Reply:
> I'd prefer not storing the fees (but the hardcoded fees) on the exchange instance – i think a better approach would be to store them on the corresponding markets... So, maybe, we need

markets are volatile so if they get reloaded the fees will be overwritten, no? and if cache the previous market in setMarkets and the reply doesn't give us all the markets we lose some fee info. So `loaded_fees` has to be a separate property, and therefore we should rename, or at least make fees point to, hardcoded fees.

Reply:
> markets are volatile so if they get reloaded the fees will be overwritten, no?

Yes, and this is exactly the intent of reloading in the first place. But it doesn't mean that the info will necessarily be lost. 

We could implement it in a way such that the fee info is either kept or reloaded — depending on what the user wants. Our end-goal is to make it convenient, so, for example, suppose that the signature of `loadMarkets` could be refined to `loadMarkets (reload, reloadTradingFees)` or `loadMarkets (reloadParams = false)` or even further... I'm not saying that it has to be done this way necessarily, i'm just saying that you can approach it in a bit more open-ended mindset :p

> `loaded_fees` has to be a separate property

I wouldn't say it **has to be** a separate property, but i'd agree that it **could be** a separate property or it **could be** a property of a market. 

Also, there's an issue with it – it works for the exchanges that have exchange-wide fees, but the current structure doesn't really fit exchanges with market-specific fees. If it is a separate property, however, then it has to be indexed by symbol. And then it starts to resemble the `.markets` property... So, I'm just thinking, maybe, we don't really need a separate property after all? You could think of `.markets` as a cache of most recent market info – loaded and reloaded on top over time. Lmk your thoughts on all of this.

We've tried to approach this problem earlier, but it's a hard one. Maybe even the hardest one, and we're just starting to approach a more or less unified vision.

Reply:
> could be a property of a market.

A reason to keep loaded fees in a seperate property is that sometimes `loadMarkets` will not load all the markets, just a subset. If the call to `load_fees` loads more markets than are available in `.markets` and we want to store them in the `.markets` property, then we have to make more markets for just fee data (and we don't know where to store them because the key they are indexed under doesn't exist in `markets_by_id`). If we just drop them then we lose fee information. 

So therefore I suggest we use `market_ids` as keys for loaded trading fees, but we can use codes (instead of currency_ids) for funding fees. In other words if you recieve `{ 'XRPXLM': 3.141 }` from an exhange and `XRPXLM` is not in `marketsById` because the market is temporarily down, we still want a place to store that information.

Maybe the best way to go about this is to store all the loaded fees in `loaded_fees` and only populate existing markets with trading fee data, and existing currencies with funding fee data. We should prioritize using the loaded fees first, but if they have not been loaded for some reason then we should use hardcoded fees as a fallback. The cool thing about using a timestamp is that it can also distinguish between the hardcoded and loaded fees (we just set the timestamp to -1 for hardcoded fees).